### **PRECSION**


![image](3.png) 

Here in the picuter, accuracy of the model is 0.85, which means that model is correct 85% of the time. For the another model too, accuracy is also 85%. Here false postive of A is > false postive of B and false negetive of A is < false negative of B.\
In this example more dangerous would be false postive meaning predicting the email spam when it is not. Thus we will select the model which has less false positive. That is we will use precision to compare the models.\
The meaning of the precision is the proportion of correctly predicted positives (true positives) to the total predicted positives (true positives + false positives).\
In our example above it would be the proportion of no of mails correctly classified as spam to the no of predicted spams.

$$
precision =\frac{TP}{TP+FP}
$$

### **RECALL**
![image](4.png)

Here in the picture, accuracy of the model is 0.90, which means that model is correct 90% of the time. For the another model too accuracy is also 90%.
In the above, if we donot detect cancer for the patient who has cancer it would be more dangerous than to detect cancer for the non cancerous patient.
Thus we will use proportion of the correctly predicted postives to compare the above models among the total real postives.
Recall is used when we are more concerned with the rate of false negatives than the false positives, as in the hospital data. Recall measures the proportion of correctly identified actual positives (true positives) out of all actual positives (true positives + false negatives).

$$
Recall=\frac{TP}{TP+FN}
$$

### **F1-SCORE**
The F1-score is a metric that combines both precision and recall into a single value by taking their harmonic mean. It is especially useful when you want to balance the trade-off between precision and recall, and when you have an uneven class distribution.

In situations where both false positives and false negatives are important, the F1-score provides a better measure of a model’s performance than accuracy alone. For example, in cases where you want to ensure that not only are your positive predictions correct (high precision), but also that you are capturing as many actual positives as possible (high recall), the F1-score gives a single metric that reflects both concerns.

The F1-score is particularly helpful when you need to find an optimal balance between precision and recall, such as in medical diagnostics or spam detection, where both types of errors can have significant consequences.

The F1-score is calculated as follows:

$$ F1 = 2 \times \frac{precision \times recall}{precision + recall} $$

It ranges from 0 to 1, where a higher F1-score indicates better model performance in terms of both precision and recall.

Using the harmonic mean we obtain the value closer to the lower value. That means the F1 score will be toward precision or recall which is lower.

### **CODE DEMONSTRATION**

In [1]:
import numpy as np
import matplotlib.pyplot as plt 
from sklearn.metrics import recall_score,precision_score,f1_score


In [3]:
import pandas as pd


In [5]:
data=pd.read_csv('C:/Users/Acer/OneDrive/Documents/MACLEARNING/machinelearning/data/advertising.csv')

In [10]:
data=data.rename(columns={'Clicked on Ad':'click','Daily Time Spent on Site':'time','Age':'age'})

In [11]:
data.head()

time  age  Area Income  Daily Internet Usage  \
0  68.95   35     61833.90                256.09   
1  80.23   31     68441.85                193.77   
2  69.47   26     59785.94                236.50   
3  74.15   29     54806.18                245.89   
4  68.37   35     73889.99                225.58   

                           Ad Topic Line            City  Male     Country  \
0     Cloned 5thgeneration orchestration     Wrightburgh     0     Tunisia   
1     Monitored national standardization       West Jodi     1       Nauru   
2       Organic bottom-line service-desk        Davidton     0  San Marino   
3  Triple-buffered reciprocal time-frame  West Terrifurt     1       Italy   
4          Robust logistical utilization    South Manuel     0     Iceland   

             Timestamp  click  
0  2016-03-27 00:53:11      0  
1  2016-04-04 01:39:02      0  
2  2016-03-13 20:35:42      0  
3  2016-01-10 02:31:19      0  
4  2016-06-03 03:36:18      0

In [12]:
X=data[['time','age']]
y=data['click']

In [13]:
from sklearn.model_selection import train_test_split

In [14]:
X_train,X_test,y_train,y_test=train_test_split(X,y,random_state=43,test_size=0.3,stratify=y)

In [18]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix

In [19]:
log=LogisticRegression()
log.fit(X_train,y_train)
y_pred1=log.predict(X_test)
tree=DecisionTreeClassifier()
tree.fit(X_train,y_train)
y_pred2=tree.predict(X_test)

In [20]:
print("for logititc regression")
print("-"*60)
cdf=pd.DataFrame(confusion_matrix(y_test,y_pred1),columns=list(range(0,2)))
print(cdf)
print('precision:',precision_score(y_test,y_pred1))
print('recall:',recall_score(y_test,y_pred1))
print('f1 score:',f1_score(y_test,y_pred1))
print("\nfor decision tree")
print("-"*60)
cdf=pd.DataFrame(confusion_matrix(y_test,y_pred2),columns=list(range(0,2)))
print(cdf)
print('precision:',precision_score(y_test,y_pred2))     
print('recall:',recall_score(y_test,y_pred2))
print('f1 score:',f1_score(y_test,y_pred2))


for logititc regression
------------------------------------------------------------
     0    1
0  136   14
1   26  124
precision: 0.8985507246376812
recall: 0.8266666666666667
f1 score: 0.8611111111111112

for decision tree
------------------------------------------------------------
     0    1
0  134   16
1   31  119
precision: 0.8814814814814815
recall: 0.7933333333333333
f1 score: 0.8350877192982457


### **MULTI CLASS PRECISON AND RECALL**

### Multiclass Precision, Recall, and F1-Score

In multiclass classification, precision, recall, and F1-score are calculated for each class individually, treating each class as the "positive" class and the rest as "negative." To summarize performance across all classes, we use averaging methods such as **macro** and **weighted**.

#### **Macro Averaging**
- **Macro** calculates the metric independently for each class and then takes the average (unweighted mean).
- Each class is treated equally, regardless of how many samples it has.
- Useful when all classes are equally important.

#### **Weighted Averaging**
- **Weighted** calculates the metric for each class, then takes the average weighted by the number of true instances for each class (support).
- Classes with more samples have a greater impact on the final score.
- Useful when class imbalance is present.

#### **Example**

Suppose you have a 3-class classification problem (classes: A, B, C) and the following confusion matrix:

|        | Pred A | Pred B | Pred C |
|--------|--------|--------|--------|
| **A**  |   40   |   2    |   3    |
| **B**  |   4    |   35   |   1    |
| **C**  |   6    |   2    |   30   |

- **Precision for A** = TP / (TP + FP) = 40 / (40 + 4 + 6) = 40 / 50 = 0.80  
- **Recall for A** = TP / (TP + FN) = 40 / (40 + 2 + 3) = 40 / 45 ≈ 0.89  
- **F1 for A** = 2 × (Precision × Recall) / (Precision + Recall) ≈ 0.84

Repeat for B and C.

##### **Macro Average**
- Macro Precision = (Precision_A + Precision_B + Precision_C) / 3
- Macro Recall = (Recall_A + Recall_B + Recall_C) / 3
- Macro F1 = (F1_A + F1_B + F1_C) / 3

##### **Weighted Average**
- Weighted Precision = (Precision_A × Support_A + Precision_B × Support_B + Precision_C × Support_C) / Total Samples
- Weighted Recall and F1 are calculated similarly.

#### **In scikit-learn**

```python
from sklearn.metrics import precision_score, recall_score, f1_score

# y_true and y_pred are your true and predicted labels
precision_macro = precision_score(y_true, y_pred, average='macro')
recall_macro = recall_score(y_true, y_pred, average='macro')
f1_macro = f1_score(y_true, y_pred, average='macro')

precision_weighted = precision_score(y_true, y_pred, average='weighted')
recall_weighted = recall_score(y_true, y_pred, average='weighted')
f1_weighted = f1_score(y_true, y_pred, average='weighted')
```

- Use `average='macro'` for macro average.
- Use `average='weighted'` for weighted average.

**Summary:**  
- **Macro**: Equal weight to each class.  
- **Weighted**: Weight by class frequency.  
- Choose based on whether you want to treat all classes equally or account for class imbalance.